core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1306,7 +1309,7 @@ class Log(@volatile private var _dir: File,
     // in an unclean manner within log.flush.start.offset.checkpoint.interval.ms. The chance of this happening is low.
     maybeHandleIOException(s"Exception while increasing log start offset for $topicPartition to $newLogStartOffset in dir ${dir.getParent}") {
       lock synchronized {
-        if (newLogStartOffset > highWatermark)
+        if (reason != SegmentCompaction && newLogStartOffset > highWatermark)
```

Comment:
I don't think it is safe for replicas to compact past the `highWatermark`. Should we remove this change and make sure that log compaction doesn't compact past the `highWatermark`.

Reply:
I talked to Jason about this a bit and since we don't clean the active segment, cleaning past the high watermark is pretty rare. But I think it makes sense to explicitly enforce this in code.

I think one other change we will need to make is to update a lot of our log cleaner tests since a large portion of them clean past the high watermark. (They don't update the high watermark at all)

Reply:
Okay. We can do this in a future PR if you like. Do you want to file an issue for this?

Reply:
Added to this PR, may still want to do separately if this has gotten too big.